### PR TITLE
Christmas task

### PR DESCRIPTION
**Pull request by VaniaToper**
1. ### Task: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/christmas-task/christmas-task-part1.md
------------
2. ### Screenshot:
image

------------
3.  ### Deploy: https://christmas-task-vaniatoper.netlify.app/
------------
4. ### Done 21.12.2021 / deadline 21.12.2021
------------
5 ### Score 190/220

- [ ] Сохранение настроек в local storage -10
- [ ] Дополнительный функционал на выбор -20

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      isFav,
+      setIsFav,
+    }}>
+      <BrowserRouter>
+        <Switch>
+          <Route path='/' exact render={() => <Home />} />
+          <Route path='/filters' render={() => <Filters data={data} />} />
+          <Route path='/game' render={() => <Game />} />
+          <Route path='/NotFound' render={() => <NotFound />} />
+          <Redirect to='/NotFound' />
```
Comment (viskosa) [inline src/App.tsx]: You also can use <Route path="*" render={() => <NotFound />} /> for catching non existing routes.
It means all routes which are not processed by previous routes will be caught by this one

File: src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import { FavContext } from './context';
+import Game from './components/GamePage/Game';
+
+const App: React.FC = () => {
+  const [isFav, setIsFav] = useState([]);
+  return (
+    <FavContext.Provider value={{
+      isFav,
+      setIsFav,
+    }}>
```
Comment (viskosa) [inline src/App.tsx]: is it works? I've thought we should pass context in array: value={[isFav, setIsFav]}

File: src/components/FiltersPage/Filters.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import React, { useContext, useMemo, useState } from 'react';
```
Comment (viskosa) [inline src/components/FiltersPage/Filters.tsx]: I would recommend to move this file to the 'pages' folder and remove FiltersPage folder here. It would be better to put all components to the folder 'Components' without dividing them into different folders like 'FiltersPage' etc.

File: src/components/FiltersPage/Filters.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import { ICards, ISliderValue } from '../../types/ITypes';
+import Input from '../UI/input/Input';
+import { FavContext } from '../../context';
+import Modal from '../UI/modal/Modal';
+
+interface IProps {
+  data: ICards[];
+}
+
+const Filters: React.FC<IProps> = ({ data }) => {
```
Comment (viskosa) [inline src/components/FiltersPage/Filters.tsx]: minor comment - you can get the FC from react at string #1 and simply use FC<IProps> here

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      </form>
+      <div className={s.parameters__selects}>
+        <Select zIndex={'5'} setType={setSelectType} valueClose={'close'}
+                valueOpen={'open'}
+                onChange={setSort}
+                options={options} title={'Sort By'} />
+        <Select zIndex={'4'} setType={setSelectType} valueClose={'reverseClose'}
+                valueOpen={'reverseOpen'}
+                onChange={setSort}
+                options={optionsReverse} title={'Sort By'} />
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: Formatting should looks like this - it is common practice to format such way:
image

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        <div className={s.parameters__sort_sliders}>
+          <Slider onChange={setYearValue} value={yearValue} min={1940}
+                  max={2020} />
+          <Slider onChange={setCountValue} value={countValue} min={1}
+                  max={12} />
+        </div>
+        <ResetButton resetSliders={{
+          year: setYearValue,
+          count: setCountValue,
+        }} resetFilters={setFilter} />
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: here as well
It is hard to read code with such formatting, please do like this:
image

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                        width={item.width} key={item.value}
+                        name={item.name} />
+            ))}
+          </div>
+        </div>
+        <div className={s.parameters__filter_item}>
+          Favorites
+          <Checkbox onChange={setFav} value={'fav'} type={'fav'} />
+        </div>
+      </form>
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: I would recommend do divide this block to three independent:
1 - block with checkboxes  and their logic 
2 - block with selects  and their logic 
3 - block with sliders and their logic

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                                          setFilter,
+                                          filter,
+                                          setSort,
+                                          setYearValue,
+                                          yearValue,
+                                          setFav,
+                                          countValue,
+                                          setCountValue,
+                                          setSelectType,
+                                        }) => {
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: It is better to use such code formatting:
image

File: src/components/UI/checkbox/Checkbox.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                                      onChange,
+                                    }) => {
+
+  return (
+    <label className={s.checkbox__container}>
+      <input onChange={(e) => onChange ? onChange(e.target.checked) : undefined}
+             defaultChecked={type !== 'fav'} value={value} type='checkbox'
+             className={s.checkbox__input} />
+      {type === 'color' && <span className={s.checkbox__checkmark_color}
+                                 style={{ background: background }} />}
```
Comment (viskosa) [inline src/components/UI/checkbox/Checkbox.tsx]: it is very hard to follow the code with such formatting. Please, do like this everywhere:
image

File: src/style-dev/nullstyle.scss [hunk 10]
```diff
@@ -0,0 +1,83 @@
+* {
```
Comment (viskosa) [inline src/style-dev/nullstyle.scss]: usually the name of such file is 'reset.scss' or 'default.scss'

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    <div className={s.parameters}>
+      <form onChange={(e) => {
+        const target = e.target as HTMLInputElement;
+        changeFilter(target.checked, target.value);
+      }} className={s.parameters__filter}>
+        <div className={s.parameters__filter_item}>
+          <span>Color</span>
+          <div className={s.parameters__filter_color}>
+            {checkBoxes.colorFilter.map(item => (
+              <Checkbox type={'color'} key={item.value} value={item.value}
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: you can pass just type='color' without curly braces here and everywhere

File: src/components/UI/checkbox/Checkbox.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      {(type === 'shape' || type === 'size') &&
+        <svg className={s.checkbox__checkmark_shape}>
+          <use style={{ transform: `scale(${width})` }}
```
Comment (viskosa) [inline src/components/UI/checkbox/Checkbox.tsx]: seems it doesn't work as you pass pixels here, it has to be scale instead.
scale(1.5)

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+  const reset = () => {
+  };
+
+
+  return (
+    <div className={s.parameters}>
+      <form onChange={(e) => {
+        const target = e.target as HTMLInputElement;
+        changeFilter(target.checked, target.value);
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: onChange handler should be moved to the separate variable

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+      type: 'reverse',
+    },
+    {
+      id: '4',
+      value: 'year',
+      title: 'By Date Reverse',
+      type: 'reverse',
+    },
+  ]);
+  const changeFilter = (checked: Boolean, value: string) => {
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: 'boolean' should be written in lowercase

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    if (!checked) return setFilter([...filter, value]);
+    return setFilter(filter.filter((param) => {
+      return param !== value;
+    }));
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: it works opposite way from it should be. When you click on green checkbox, it should take only green toys and render them. If you click green and red checkboxes, it should render only green and red toys

File: src/components/FiltersPage/FiltersBlock.tsx [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+                                          filter,
+                                          setSort,
+                                          setYearValue,
+                                          yearValue,
+                                          setFav,
+                                          countValue,
+                                          setCountValue,
+                                          setSelectType,
+                                        }) => {
+  const [checkBoxes] = useState({
```
Comment (viskosa) [inline src/components/FiltersPage/FiltersBlock.tsx]: it's not neccesary to use useState here as you don't use setState function. Just const would be ok

const checkBoxes = {colorFilter: [...], shapeFilter: [...], ...}

File: src/components/UI/checkbox/Checkbox.tsx (hunk 9 above)
Comment (viskosa) [reply src/components/UI/checkbox/Checkbox.tsx]: Even better would be move the rendering into separate variable and make decision through switch case. It is more readable

image

File: src/components/FiltersPage/Filters.tsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+               onChange={setSearchQuery} />
+        {sortCard.length
+          ? sortCard.map((card, index) => (
+            <Card value={fav} key={index} onChange={setFav} card={card} />))
+          : <span>Not Found</span>
+        }
+      </div>
+    </div>
+  );
+};
```
Comment (viskosa) [inline src/components/FiltersPage/Filters.tsx]: Do formatting like this, please:
image

File: src/components/FiltersPage/Card.tsx [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+                                               shape,
+                                               color,
+                                               size,
+                                               favorite,
+                                               name,
+                                               num,
+                                             },
+                                             onChange,
+                                             value,
+                                           }) => {
```
Comment (viskosa) [inline src/components/FiltersPage/Card.tsx]: fix the formatting here please. I left examples below

File: src/components/FiltersPage/Card.tsx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+  const { isFav, setIsFav } = useContext(FavContext);
+  const setFavCards = (checked: Boolean, num: string) => {
```
Comment (viskosa) [inline src/components/FiltersPage/Card.tsx]: boolean should be in lowercase everywhere

File: src/components/UI/input/Input.tsx [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+const Input: React.FC<IProps> = ({
+                                   value,
+                                   onChange,
+                                   placeholder,
+                                   autoFocus,
+                                 }) => {
+  return (
+    <form className={s.search_box}>
+      <input placeholder={placeholder} autoFocus={autoFocus} className={s.input}
+             onChange={(e)=>onChange(e.target.value)} value={value} type='text' />
```
Comment (viskosa) [inline src/components/UI/input/Input.tsx]: fix formatting please

File: src/components/UI/select/Select.tsx [hunk 21]
```diff
@@ -0,0 +1,64 @@
+import React from 'react';
+import s from './Select.module.scss';
+import { IOptions } from '../../../types/ITypes';
+
+interface IProps {
+  title: string;
+  options: IOptions[];
+  onChange: React.Dispatch<React.SetStateAction<string>>;
```
Comment (viskosa) [inline src/components/UI/select/Select.tsx:8]: I think far all such handlers returning void will be enough
onChange: () => void
onBlaBla: () => void

File: src/components/UI/select/Select.tsx [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+
+          </ul>
+          <label className={s.select__expand_label}
+                 htmlFor={valueOpen} />
+        </li>
+      </ul>
+    </form>
+  );
+};
```
Comment (viskosa) [inline src/components/UI/select/Select.tsx:71]: Formatting should be fixed

File: src/components/UI/select/Select.tsx [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+                 name='tickets_type__entrance' id={valueOpen} />
+          <label className={s.select__close_label}
+                 htmlFor={valueClose} />
+          <ul className={s.select__options}>
+            {options.map((option) => (
+              <li key={option.id} className={s.select__option}>
+                <input onChange={e => {
+                  onChange(e.target.value);
+                  setType(option.type);
+                }}
```
Comment (viskosa) [inline src/components/UI/select/Select.tsx]: complex handlers should be stored in separate constants

**Review by viskosa — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 2 above)
Comment (VaniaToper) [reply src/App.tsx]: actually works :D

**Review by NProtsenko — APPROVED** [review]
